accepted/future-releases/static-extension-methods/feature-specification.md

```diff
@@ -46,214 +46,205 @@ extension MyFancyList<T> on List<T> {
 
-```dart
-extension SuperList<T> on List<T> {
-  R foldRight<R>(R base, R combine(T element, R accumulator)) {
-    for (int i = this.length - 1; i >= 0; i--) {
-      base = combine(this[i], base);
-    }
-    return base;
-  }
-}
-...
-  List<String> strings = ...;
-  int count(String string, int length) => length + string.length;
-  ...
-  var length = strings.foldRight(0, count);
-```
+If `E` is the single most specific accessible and applicable extension for a member invocation *i* with target expression `e`, which would then be a compile-time error by itself,  then we treat the target expression as if it was the extension application of `E` to `e`, and if `E` is generic, also providing the type arguments inferred for `E` in checking that it was applicable. This makes the member invocation equivalent to an explicit extension member invocation. This happens even if the *name* `E` is not accessible, so this is not a syntactic rewrite.
 
-Here the inference occurs just as if the extension had been declared like:
+Implicit extension member invocation can also apply to individual *cascade* invocations. A cascade is treated as if each cascade section was a separate member invocation on an expression with the same value as the cascade receiver expression (the one before the first `..`). This means that a cascade like `o..foo()..bar()` may perform an implicit extension member invocation on `o` for `foo()` and a normal invocation on `o` for `bar()`. There is no way to specify the corresponding explicit member invocation, so this is again not a syntactic rewrite.
 
-```dart
-class SuperList<T> {
-  final List<T> $this;
-  SuperList(this.$this);
-  R foldRight<R>(R base, R Function(T, R) combine) { ... }
-}
-```
+##### Accessibility
 
-and it was invoked as `SuperList(strings).foldRight(0, count)`.
+An extension is *accessible* for an expression if it has a name which is not private to a different library and the extension declaration is declared or imported into a scope which is a (recursive) parent scope of the lexical scope of the expression. That currently means imported into the default import scope of the current library, or declared at the top-level of the library itself.
 
-Or, alternatively, like the extension method had been declared like:
+You can *avoid* making the extension in-scope for a library by either not importing any library exporting the extension, importing such a library and hiding the extension using `hide` or `show`, or importing such a library only with a prefix.
 
-```dart
-R Function<R>(R, R Function(T, R)) SuperList$foldRight<T>(List<T> $this) =>
-    <R>(R base, R Function(T, R) combine) { ... };
-```
+An extension *is* in scope if the name is *shadowed* by another declaration (a class or local variable with the same name shadowing a top-level or imported declaration, a top-level declaration shadowing an imported extension, or a non-platform import shadowing a platform import).
```

Comment:
This assigns a completely new meaning to the phrase `in scope`. We will then need to revisit all locations where that phrase is used (the spec has 206 occurrences), and make sure that they still mean the right thing, or that they use a different phrase. Maybe use a different phrase for the new concept?

Reply:
No longer uses "in scope".